### PR TITLE
Add support for customHandler section in host.json

### LINKED ISSUE #5547: Add profile based startup options for httpWorker
cc @anthonychu

To configure debug options for HttpWorker, arguments need to passed in host.json as detailed here https://github.com/Azure/azure-functions-host/issues/5463

Profile based start up options would help to differentiate between dev and prod environments.
Another option is to introduce a start up environment variable.

### LINKED ISSUE #5615: [HttpWorker]Add configuration flag in opt-out of http trigger request forward
cc @fabiocav @mattchenderson 

Currently, if function has only httpTrigger functions host acts as a proxy forwards request to and from HttpWorker. https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs#L50

If using HttpTrigger with other bindings, request content to HttpWorker  has the format defined here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Http/HttpScriptInvocationContext.cs

Need to figure out if we should have two models or just one.

### LINKED ISSUE #5692: [HttpWorker]Expand environment variables specified in startup command
Some startup commands require `FUNCTIONS_HTTPWORKER_PORT` to be passed to it. Couldn't find a way to do this. The following doesn't seem to work:

```json
"httpWorker": {
	"description": {
		"defaultExecutablePath": "php",
		"defaultWorkerPath": "worker.php",
		"arguments": [
			"-S",
			"0.0.0.0:$FUNCTIONS_HTTPWORKER_PORT"
		]
	}
}
```

### LINKED ISSUE #5882: Custom handlers - configure working directory for worker process
Some applications are happier when you can configure the working directory in which it runs.

For instance:

```json
{
    "version": "2.0",
    "httpWorker": {
        "description": {
            "defaultExecutablePath": "node",
            "defaultWorkerPath": "myapp/server.js"
        }
    }
}
```

Would be great if we could do something like the following instead:

```json
{
    "version": "2.0",
    "httpWorker": {
        "description": {
            "defaultExecutablePath": "node",
            "defaultWorkerPath": "server.js",
            "defaultWorkingDirectory": "myapp"
        }
    }
}
```

### LINKED ISSUE #5921: Custom handlers - Forward path from original request
When proxying HTTP requests, it doesn't appear the worker is able to access the route from the original request.

For instance, given this host.json and function.json to create a function that triggers on every route:

```json
{
  "version": "2.0",
  "httpWorker": {
    "description": {
      "defaultExecutablePath": "node",
      "defaultWorkerPath": "server.js"
    }
  },
  "extensions": {
    "http": {
      "routePrefix": ""
    }
  }
}
```

```json
{
  "bindings": [
    {
      "type": "httpTrigger",
      "authLevel": "anonymous",
      "direction": "in",
      "name": "req",
      "route": "{*restofpath}"
    },
    {
      "type": "http",
      "direction": "out",
      "name": "$return"
    }
  ]
}
```

It doesn't look like the worker is able to access the original route (`restofpath`). Perhaps this can be made available as an HTTP header?

### LINKED ISSUE #6080: Custom Handlers - Change "httpWorker" section to "custom" in host.json
As discussed, let's change the `httpWorker` section to `custom` (or should it be `customHandler`?):

```json
{
    "version": "2.0",
    "customHandler": {
        "type": "http",

    }
}
```

If possible, it'd be great to support both the old and new sections for a couple of releases and output a deprecation warning at startup. If both sections are in host.json, throw an error.

/cc @pragnagopa @fabiocav @yojagad

### PR DESCRIPTION
**Pull request by pragnagopa**
This PR adds support to specify custom handler section in host.json as discussed in #6080

- Here is an example `host.json` that specifies `workingDirectory` 
```json
{
	"version": "2.0",
	"extensionBundle": {
		"id": "Microsoft.Azure.Functions.ExtensionBundle",
		"version": "[1.*, 2.0.0)"
	},
	"customHandler": {
		"description": {
			"defaultExecutablePath": "dotnet",
			"workingDirectory": "CSharpCustomHandlers/bin/Debug/netcoreapp2.2",
			"arguments": [ "CSharpCustomHandlers.dll" ]
		},
                "enableForwardingHttpRequest": true
	}
}
```
- Environment variables can be specified as part of any element in `description` following the format 
`%EnvVarName%`. Here is an example
```json
{
	"version": "2.0",
	"customHandler": {
		"type": "http",
		"enableForwardingHttpRequest": false,
		"description": {
			"defaultExecutablePath": "%GoWorkerEnv%",
			"arguments": [
				"%FUNCTIONS_CUSTOMHANDLER_PORT%"
			]
		}
	}
}
```

- Each element in `description` can also be overridden via EnvVars/AppSettings. Here is an example of AppSetting / Environment variable name to overrride arguments element
`AzureFunctionsJobHost:customHandler:description:arguments`

Fixes #5692, Fixes #5921, Fixes #5547, Fixes #6080, Fixes #5882, Fixes #5615

Note: httpWorker section will be deprecated sometime later. For now, it is still supported.

### CODE REVIEW

File: src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs]: @anthonychu - This will now throw explicit file not found exception if file specified in `DefaultWorkerPath` or `DefaultExecutablePath` does not exist

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerDescription.cs
Comment (yojagad) [inline src/WebJobs.Script/Workers/ProcessManagement/WorkerDescription.cs:79]: Would the RpcWorkerDescription need this too? Or is this HttpWorker specific?

**Review by yojagad — DISMISSED** [review]
:shipit:

**Comment by anthonychu** [comment]
Can we please hold off on this while we discuss #6020? Thanks.

**Comment by pragnagopa** [comment]
> Can we please hold off on this while we discuss #6020? Thanks.

Yes. I will hold off. Added comments in referenced issue. This PR would still be needed to support env vars in the existing config.

File: src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs:27]: Is `WorkerArgument` required to be `List`? Can we change this to an `IEnumerable` and use `Enumerable.Empty<string>()` here instead?

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: There's a bit of duplication here and in the file below, causing us to loop twice and create additional collections. Could we consolidate this in a single place? 

Also, the name `Arguments` and `WorkerArguments` are confusing; I vaguely remember talking about that but don't recall where we landed. What is the distinction (runtime/worker arguments)? Would be good to have more clarity in the name

**Review by fabiocav — COMMENTED** [review]
Couple of comments from a brief look.

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs
Comment (anthonychu) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: There's some further discussion here: https://github.com/Azure/azure-functions-host/issues/6020#issuecomment-631795093

Would like to see how much we can simplify this. Want to take the opportunity to make changes while we are changing from `httpWorker` to `customHandler`.

Comment (fabiocav) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: Yes, I think the can become simpler. Will take a look and catch up on the conversation there. Thanks!

Comment (fabiocav) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: Took a look at the comment and it is what I remembered. The fact that I was confused again is a sign that we probably need to come up with a better name/approach

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: Thank you both for providing feedback. I will hold off this PR in favor of https://github.com/Azure/azure-functions-host/issues/6080

File: src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs:27]: The list of arguments is not static and is being updated later to allow runtime to append arguments if needed. Updated to `IList`

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: This PR is now includes changes discussed in #6080. Leaving `Arguments` and `WorkerArguments` as advanced options. Updated PR description with an example host.json. Please take a look.

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs:8]: @anthonychu - should we leave out the `Type`? We can introduce `Type`, in future if we ever decide to expand customHandlers to any other tranport protocol. We can then default to `http` if type is not sepcified.

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs]: @anthonychu / @yojagad / @fabiocav - Introducing `EnableHttpRequestForward` flag to opt out of proxying http trigger request. Setting this to true by default will make getting started with customHandlers easier. If the user decides to opt-out this flag can be set to `false`

Let me know what you think and also if you have better suggestions for the name.

Note: This flag only applies to functions that only http in and out.

File: src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs:18]: @anthonychu - Take a look at the environment variable names.

File: src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/HttpWorkerDescription.cs:27]: The list of arguments is not static and is being updated later to allow runtime to append arguments if needed. Updated to `IList`

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs:54]: `WorkingDirectory` default to function app root if not specified.

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs]: This error will now include filename for the process.

**Review by yojagad — DISMISSED** [review]
:shipit:

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (anthonychu) [reply src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs:8]: I'll defer to you and @fabiocav on this one. I'm okay with it being required, optional, or omitted altogether.

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (anthonychu) [reply src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs]: I think I would prefer this to default to `false` but I can be convinced otherwise. My main reasons are:

* When it's off, behavior is consistent no matter what triggers and bindings are in the functions.
* It's easier to document.
  - The main documentation can focus on a single way of doing things.
  - Then we can have a section about HTTP request forwarding: how to enable and how it works. This section could potentially contain info on how to forward all routes to handler.

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptionsSetup.cs
Comment (anthonychu) [inline src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptionsSetup.cs]:
```suggestion
        private static List<string> GetArgumentList(IConfigurationSection httpWorkerSection, string argumentSectionName)
        {
            var argumentsSection = httpWorkerSection.GetSection(argumentSectionName);
```

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs
Comment (anthonychu) [inline src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs:177]: `System.Uri.AbsolutePath` doesn't contain the query string. If forwarding the request to the handler, I think we want to include the query string as well.

```suggestion
                pathValue = httpRequestMessage.RequestUri.PathAndQuery;
```

File: src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs
Comment (anthonychu) [reply src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs:18]: Will we continue to populate `FUNCTIONS_HTTPWORKER_PORT` and `FUNCTIONS_CUSTOMHANDLER_PORT`? (either until we remove the `httpWorker` section or maybe forever just to avoid breaking existing apps)

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs
Comment (anthonychu) [reply src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs]: Is it possible to log the entire startup command (as much info as possible, maybe even working directory)?

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs]: Ok makes sense. Will update.

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs]: Will do.

File: src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/HttpWorkerConstants.cs:18]: we can leave them.

File: src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs:177]: Query string is appended here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Http/DefaultHttpWorkerService.cs#L160

**Comment by pragnagopa** [comment]
@yojagad - can you please take quick a look again as I had to resolve few merge conflicts

File: src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Http/Configuration/HttpWorkerOptions.cs:8]: We can leave this  out of documentation. Defaults to `http` for now.

**Comment by pragnagopa** [comment]
@fabiocav  - can you please take a look at the config `customhandler` options ? It would be good to to get this PR in for the next release.

File: src/WebJobs.Script/Config/HostJsonFileConfigurationSource.cs
Comment (yojagad) [inline src/WebJobs.Script/Config/HostJsonFileConfigurationSource.cs]: >customHandler [](start = 17, length = 13)

aren't we supporting both for sometime  ?

Comment (pragnagopa) [reply src/WebJobs.Script/Config/HostJsonFileConfigurationSource.cs]: You are right. Will be good to leave this here. No need to add metric but have having host.json will help us differentiate. Will update

**Comment by yojagad** [comment]
That was the only comment I had. Rest looks good!

**Review by yojagad — APPROVED** [review]

**Pull request merged**
